JPA Map storage listener should handle optimistic locking for deleting entities

### Describe the bug

When deleting an entity, the following exception can occur:

``` 
Caused by: java.lang.IllegalArgumentException: entity not in the persistence context
	at org.hibernate.internal.SessionImpl.lock(SessionImpl.java:3613)
	at org.hibernate.internal.SessionImpl.lock(SessionImpl.java:3604)
	at org.keycloak.models.map.storage.jpa.JpaChildEntityListener.checkRoot(JpaChildEntityListener.java:49)
	at org.keycloak.models.map.storage.jpa.JpaChildEntityListener.onPreDelete(JpaChildEntityListener.java:61)
	at org.hibernate.action.internal.EntityDeleteAction.preDelete(EntityDeleteAction.java:137)
	at org.hibernate.action.internal.EntityDeleteAction.execute(EntityDeleteAction.java:78)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:478)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:356)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1472)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:512)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3310)
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2506)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:447)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:178)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.access$300(JdbcResourceLocalTransactionCoordinatorImpl.java:39)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:271)
	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:104)
	at org.keycloak.connections.jpa.JpaKeycloakTransaction.commit(JpaKeycloakTransaction.java:49)
```

### Version

17.0.0-SNAPSHOT

### Expected behavior

Tests like ClientSearchTest should complete successfully.

### Actual behavior

An exception is thrown.

### Anything else?

This was first reported by @vramik when running tests locally.

--- a/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/hibernate/listeners/JpaOptimisticLockingListener.java
+++ b/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/hibernate/listeners/JpaOptimisticLockingListener.java
@@ -28,6 +28,7 @@ import org.hibernate.event.spi.PreUpdateEventListener;
 import org.keycloak.models.map.storage.jpa.JpaChildEntity;
 
 import javax.persistence.LockModeType;
+import java.util.Objects;
 
 /**
  * Listen on changes on child entities and forces an optimistic locking increment on the topmost parent aka root.
@@ -46,8 +47,15 @@ public class JpaOptimisticLockingListener implements PreInsertEventListener, Pre
             Object root = entity;
             while (root instanceof JpaChildEntity) {
                 root = ((JpaChildEntity<?>) entity).getParent();
+                Objects.requireNonNull(root, "children must always return their parent, never null");
             }
-            session.lock(root, LockModeType.OPTIMISTIC_FORCE_INCREMENT);
+
+            // a session would not contain the entity if it has been deleted
+            // if the entity has been deleted JPA would throw an IllegalArgumentException with the message
+            // "entity not in the persistence context".
+            if (session.contains(root)) {
+                session.lock(root, LockModeType.OPTIMISTIC_FORCE_INCREMENT);
+            }
         }
     }
 

